splink/comparison_level_composition.py

```diff
@@ -0,0 +1,262 @@
+from __future__ import annotations
+
+from typing import Iterable
+
+from splink.comparison_level import ComparisonLevel
+
+
+def cl_and(
+    *clls: ComparisonLevel | dict,
+    label_for_charts=None,
+    m_probability=None,
+    is_null_level=None,
+) -> ComparisonLevel:
+
+    """Merge ComparisonLevels using logical "AND".
+
+    Merge multiple ComparisonLevels into a single ComparisonLevel by
+    merging their SQL conditions using a logical "AND".
+
+    By default, we generate a new ``label_for_charts`` for the new ComparisonLevel.
+    You can override this, and any other ComparisonLevel attributes, by passing
+    them as keyword arguments.
+
+    Args:
+        *clls (ComparisonLevel | dict): ComparisonLevels or comparison
+            level dictionaries to merge
+        label_for_charts (str, optional): A label for this comparson level,
+            which will appear on charts as a reminder of what the level represents.
+            Defaults to a composition of - `label_1 AND label_2`
+        m_probability (float, optional): Starting value for m probability.
+            Defaults to None.
+        is_null_level (bool, optional): If true, m and u values will not be
+            estimated and instead the match weight will be zero for this column.
+            Defaults to None.
+
+    Examples:
+        >>> # Simple null level composition with an `AND` clause
+        >>> import splink.duckdb.duckdb_comparison_level_library as cll
+        >>> cll.cl_and(cll.null_level("first_name"), cll.null_level("surname"))
+
+        >>> # Composing a levenshtein level with a custom `contains` level
+        >>> import splink.duckdb.duckdb_comparison_level_library as cll
+        >>> misspelling = cll.levenshtein_level("name", 1)
+        >>> contains = {
+        >>>     "sql_condition": "(contains(name_l, name_r) OR " \
+        >>>     "contains(name_r, name_l))"
+        >>> }
+        >>> merged = cl_and(misspelling, contains, label_for_charts="Spelling error")
+        >>> merged.as_dict()
+        >>> {
+        >>>    'sql_condition': '(levenshtein("name_l", "name_r") <= 1) ' \
+        >>>    'AND ((contains(name_l, name_r) OR contains(name_r, name_l)))',
+        >>>    'label_for_charts': 'Spelling error'
+        >>> }
+
+    Returns:
+        ComparisonLevel: A new ComparisonLevel with the merged
+            SQL condition
+    """
+
+    return _cl_merge(
+        *clls,
+        clause="AND",
+        label_for_charts=label_for_charts,
+        m_probability=m_probability,
+        is_null_level=is_null_level,
+    )
+
+
+def cl_or(
+    *clls: ComparisonLevel | dict,
+    label_for_charts=None,
+    m_probability=None,
```

Comment:
I like how you made these explicit args instead of kwargs.

Including `label_for_charts` and `is_null_level` makes sense to me, but why `m_probability` and not all the other more esoteric stuff like `tf_adjustment_weight`? I personally have never manually set ``m_probability` so I might not understand why you would want to.

Reply:
Term frequencies need some work before we can include them here. At present you can only tf adjust a single column, not multiple.

`m_probability` we do use internally. It's also present in all of our pre-built [comparison levels](https://github.com/moj-analytical-services/splink/blob/master/splink/comparison_level_library.py), so I think it should be included here too.